### PR TITLE
Add function to resolve caseless path to caseful variant

### PR DESCRIPTION
**Pull request by selaux**
This should extract a common functionality used to find a directory case insensitively. Basically now all user editable paths should be canonicalized and made case insensitive with this commit. It also seems to fix one of the issues in #886 where the launcher mishandles spaces.

### CODE REVIEW

**Comment by flaviojs** [comment]
You must take extra care with `caseless=false`tests because the case sensitivity of the filesystem can't be predicted. With symbolic links, even just part of the path can have a different case sensitivity.

Example: `t!(temp_dir.path(), "FOO", false, <expected result>);`
Here the expected result can be either true or false, it is impossible to predict.

As for the strategy of always comparing with Nfc, I'm unsure... do all filesystems treat names with different umlaut ö variations as the same name?
They probably don't, and that would make `resolve_existing_components` wrong because it can resolve to the wrong path.

As for #886, you should note that 0.16.1 is being used and the launcher changed a lot since then.
The space issue might already be fixed in master.

**Comment by selaux** [comment]
True, I missed the case-sensitivity of the OSs in the tests :confused:.

Regarding always comparing with NFC, I would argue that it should always work the same, regardless of case-sensitivity. So either we always normalize to NFC (in both caseful and caseless matching) or never. Matching with NFC only in the caseless case is confusing in my oppinion. Additionally a user having two existing paths with different unicode variants seems a very unlikely edge case to me. But I'm open for suggestions.

**Comment by flaviojs** [comment]
It only seems weird if you view paths as true unicode strings.
I view them as raw data that can even include invalid unicode and the `\0` character.
I'm not sure how much we support when coming from the C/C++ side, that is something for when filesystem stuff is rustified.

If we ask for a case sensitive path we are saying we want an **exact** match to the argument, so we should make raw comparisons.

If we ask for a case insensitive path we are saying we want a match according to some case folding rules. How we actually apply case folding (in NFC form) is an implementation detail.
With multiple matches, the best possible match is a raw match.

Ideally we would ask a case insensitive filesystem for the correct path components, but we can't.
Currently symbolic link aren't being considered. I don't know if we should follow them here.

**Comment by flaviojs** [comment]
Hmm, I think imposing NFC is fine if we're taking about a virtual filesystem where we control the rules (like the content manager).

But this function is in the filesystem module (fs) so it should respect the filesystem as much as possible.

**Comment by selaux** [comment]
Okay I will change the PR then to the following:

- Dont use NFC when not matching caseless paths
- Follow links when resolving paths

**Comment by flaviojs** [comment]
Not following links might be ok.
Even though I remember reading several times about following links, I don't remember any specific reason why it's needed. :/

**Comment by flaviojs** [comment]
Closed by mistake when I used push instead of pull. >.<

**Comment by flaviojs** [comment]
I'll try to continue this PR.

**Comment by selaux** [comment]
Thanks. Sorry, currently I have only little time for this project.

**Comment by flaviojs** [comment]
With the current implementation the name `resolve_existing_components` looks misleading to me.
Maybe rename it to `join_and_resolve_caseless`?

Note that it does not canonicalize, it only normalizes path separators, everything else is maintained.

File: rust/stracciatella_c_api/src/c/misc.rs [hunk 1]
```diff
@@ -232,28 +219,59 @@ mod tests {
...
-        t!("FOO", true, true);
+
+        t!(temp_dir.path(), "baz", false, false);
+        t!(temp_dir.path(), "baz", true, false);
+
+        t!(temp_dir.path(), "foo", false, true);
+        t!(temp_dir.path(), "foo", true, true);
+        t!(temp_dir.path(), "FOO", true, true);
+
+        t!(temp_dir.path(), "foo/bar", false, true);
```
Comment (selaux) [inline rust/stracciatella_c_api/src/c/misc.rs]: Does it make sense to add a test where the base dir is `temp_dir.path().join("Foo")` to test whether the base path is resolved case insensitively as well?

**Comment by selaux** [comment]
I cant approve since I openened the PR, but other than the one comment this looks good to me.

File: rust/stracciatella_c_api/src/c/misc.rs (hunk 1 above)
Comment (flaviojs) [reply rust/stracciatella_c_api/src/c/misc.rs]: The result will depend on the case sensitivity of the filesystem, so it can't be tested.

**Pull request merged**